### DISCUSSION TITLE
Are multiple calls to LoggingFactory.configure safe, specifically the DefaultLoggingFactory implementation

### DISCUSSION THREAD

**Discussion by dennyac** [dropwizard/dropwizard#5932] [Q&A] [answered]
Hi, 

We have a use case where we want to add filters (FilterFactory) across all logging appenders and add a custom appender via a dropwizard bundle. We could do it manually via config, but want to avoid scenarios, where developers forget to add filters when new appenders are added.

When we modify appenders in the DefaultLoggingFactory, we have to call [DefaultLoggingFactory.configure](https://github.com/dropwizard/dropwizard/blob/release/2.1.x/dropwizard-logging/src/main/java/io/dropwizard/logging/DefaultLoggingFactory.java#L123) to pick up the modifications in DefaultLoggingFactory. **Wanted to know if it is safe to call DefaultLoggingFactory.configure multiple times as it is already invoked earlier during application initialization**. Seems to work, but wanted to know if it is safe to call multiple times.

```
public class AnalyticsBundle implements ConfiguredBundle<Configuration> {
    private final Bootstrap<?> bootstrap;
    private static final Set<String> EVENT_LOGGERS = Collections.singleton("AnalyticsLogger");
    private final Function<ILoggingEvent, FilterReply> onlyAllowEvents = e -> EVENT_LOGGERS.contains(e.getLoggerName()) ? FilterReply.NEUTRAL : FilterReply.DENY;
    private final Function<ILoggingEvent, FilterReply> removeEvents = e -> !EVENT_LOGGERS.contains(e.getLoggerName()) ? FilterReply.NEUTRAL : FilterReply.DENY;


    public AnalyticsBundle(Bootstrap<?> bootstrap) {
        this.bootstrap = bootstrap;
    }

    @Override
    public void run(Configuration configuration, Environment environment) {
        if(!DefaultLoggingFactory.class.isInstance(configuration.getLoggingFactory())) {
            throw new RuntimeException("AnalyticsBundle requires default dropwizard logging.");
        }
        DefaultLoggingFactory defaultLoggingFactory = (DefaultLoggingFactory) configuration.getLoggingFactory();
        defaultLoggingFactory.getAppenders().forEach(a -> {
            AbstractAppenderFactory<ILoggingEvent> abstractAppenderFactory = ((AbstractAppenderFactory<ILoggingEvent>) a);
            abstractAppenderFactory.setFilterFactories(Stream
                    .concat(abstractAppenderFactory.getFilterFactories().stream(), Stream.of(new LoggingEventFilterFactory(removeEvents))).collect(Collectors.toList()));
        });
        FileAppenderFactory<ILoggingEvent> fileAppenderFactory = new FileAppenderFactory<>();
        fileAppenderFactory.setCurrentLogFilename("logs/events.events");
        fileAppenderFactory.setArchive(false);
        fileAppenderFactory.setFilterFactories(Collections.singletonList(new LoggingEventFilterFactory(onlyAllowEvents)));
        fileAppenderFactory.setTimeZone(TimeZone.getTimeZone("UTC"));
        fileAppenderFactory.setLogFormat("%msg%n");
        defaultLoggingFactory.setAppenders(Stream.concat(defaultLoggingFactory.getAppenders().stream(),Stream.of(fileAppenderFactory) ).collect(Collectors.toList()));
        configuration.getLoggingFactory().configure(bootstrap.getMetricRegistry(), bootstrap.getApplication().getName());
    }
}

```

**Comment by zUniQueX** [comment, marked as answer]
The `configureLoggers(...)` method calls `LoggerContext.reset()`, which resets Logback's current configuration and detaches all current appenders. So if you re-add the appenders, they are newly registered and this shouldn't cause problems. Note: log messages which occur while reconfiguring will be swallowed.

However, I'm not sure, if JMX will break when you're doing this.

**Reply by dennyac** [reply]
@zUniQueX Thanks for the quick response. Is there any way to verify if JMX is broken?

I can see the logs being written to the appropriate appenders after modifying them. So I'm assuming that it works. Haven't run into any issues yet. :D

**Reply by zUniQueX** [reply]
I've now taken a deeper look at the implementation and can't see any issues _at the moment_.

This refers to Dropwizard 2.1.2 and Logback 1.2.11.

**Reply by dennyac** [reply]
@zUniQueX - Thanks for taking the time to look deeper into the implementation!
